metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+        dashboards: List[Dashboard] = [
+            Dashboard(
+                id=instance.get("id"),
+                isReadOnly=instance.get("isReadOnly"),
+                displayName=instance.get("displayName"),
+                embedUrl=instance.get("embedUrl"),
+                webUrl=instance.get("webUrl"),
+                workspace_id=workspace.id,
+                workspace_name=workspace.name,
+                tiles=[],
+                users=[],
+                tags=[],
+            )
+            for instance in dashboards_dict
+            if instance is not None
+        ]
+
+        return dashboards
+
+    def get_groups(self) -> List[dict]:
+        group_endpoint = self.get_groups_endpoint()
+        params: dict = {"$top": self.TOP, "$skip": 0, "$filter": "type eq 'Workspace'"}
+
+        def fetch_page(page_number: int) -> dict:
+            params["$skip"] = self.TOP * page_number
+            logger.debug(f"Query parameters = {params}")
+            response = self._request_session.get(
+                group_endpoint,
+                headers={Constant.Authorization: self.get_access_token()},
+                params=params,
+            )
+            response.raise_for_status()
+            return response.json()
+
+        # Hit PowerBi
+        logger.info(f"Request to groups endpoint URL={group_endpoint}")
+        zeroth_page = fetch_page(0)
+        logger.debug(f"Page 0 = {zeroth_page}")
+        if zeroth_page.get("@odata.count") is None:
+            raise ValueError(
+                "Required field @odata.count is missing. Not able to determine number of pages to fetch"
+            )
+
+        number_of_items = zeroth_page["@odata.count"]
+        number_of_pages = math.ceil(number_of_items / self.TOP)
+        output: List[dict] = zeroth_page["value"]
+        for page in range(
+            1, number_of_pages
+        ):  # start from 1 as 0th index already fetched
+            page_response = fetch_page(page)
+            if len(page_response["value"]) == 0:
+                break
+
+            logger.debug(f"Page {page} = {zeroth_page}")
+
+            output.extend(page_response["value"])
+
+        return output
+
+    def get_reports(
+        self, workspace: Workspace, _filter: Optional[str] = None
+    ) -> List[Report]:
+        reports_endpoint = self.get_reports_endpoint(workspace)
+        # Hit PowerBi
+        logger.info(f"Request to report URL={reports_endpoint}")
+        params: Optional[dict] = None
+        if _filter is not None:
+            params = {"$filter": _filter}
+
+        response = self._request_session.get(
+            reports_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+            params=params,
+        )
+        response.raise_for_status()
+        response_dict = response.json()
+        reports: List[Report] = [
+            Report(
+                id=raw_instance["id"],
+                name=raw_instance.get("name"),
```

Comment:
Are all of these operations safe? SHould we verify that raw_instance has the fields that we require? (Or use defaults)